File: DSA /Unit 05/Static/Q1.cpp
Single File Programming Question
Problem Statement



Ravi is building a basic hash table to manage student roll numbers for quick lookup. He decides to use Linear Probing to handle collisions.



Implement a hash table using linear probing where:

The hash function is: index = roll_number % table_size
On collision, check subsequent indexes (i+1, i+2, ...) until an empty slot is found.


You need to:

Insert a list of n student roll numbers into the hash table.
Print the final state of the hash table. If a slot is empty, print -1.
Input format :
The first line of the input contains two integers n and table_size, where n is the number of roll numbers to be inserted, and table_size is the size of the hash table.

The second line contains n space-separated integers representing the roll numbers to insert into the hash table.

Output format :
The output should print a single line with table_size space-separated integers representing the final state of the hash table after all insertions.

If any slot remains unoccupied, it should be represented as -1.



Refer to the sample output for formatting specifications.

Code constraints :
The given test cases fall under the following constraints:

1 ≤ n ≤ table_size

1 ≤ table_size ≤ 100

1 ≤ roll_number ≤ 1000

Sample test cases :
Input 1 :
4 7
50 700 76 85
Output 1 :
700 50 85 -1 -1 -1 76 
Input 2 :
3 5
11 22 33
Output 2 :
-1 11 22 33 -1



  



#include <iostream>
using namespace std;
#define MAX 100

void initializeTable(int table[], int table_size) {
    for (int i = 0; i < table_size; i++) {
        table[i] = -1;
    }
}

void insertIntoHashTable(int table[], int table_size, int arr[], int n) {
    for (int i = 0; i < n; i++) {
        int key = arr[i];
        int index = key % table_size;

        while (table[index] != -1) {
            index = (index + 1) % table_size;
        }
        table[index] = key;
    }
}

void printTable(int table[], int table_size) {
    for (int i = 0; i < table_size; i++) {
        cout << table[i] << " ";
    }
}

int main() {
    int n, table_size;
    cin >> n >> table_size;

    int arr[MAX];
    int table[MAX];

    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    initializeTable(table, table_size);
    insertIntoHashTable(table, table_size, arr, n);
    printTable(table, table_size);

    return 0;
}

File: Java/Unit 03/CRPC/Q1.java
Single File Programming Question
Problem Statement



You are developing a program to help analyze scores in a competition by determining the highest value from a given set of numbers. The program should allow the user to input the total number of elements and then enter the values one by one. To achieve this, you need to implement a class structure using inheritance. 



Create a base class that provides a method to find the maximum value in an array, and extend this class to a subclass that manages the input data. Additionally, override the toString() method to display the array contents in a readable format and the equals() method to compare two sets of data. The program should then compute and print the highest value from the input set.

Input format :
The first line of input consists of an interger 'n' representing the number of elements in the array.

The second line of input consists of n space-separated integers representing the array elements.

Output format :
The output prints an integer representing the maximum element in the given array.



Refer to the sample output for formatting specifications.

Code constraints :
The given testcases fall under the following constraints:

0 ≤ n ≤ 100

1 ≤ array elements ≤ 100

Sample test cases :
Input 1 :
5
1 6 9 8 7
Output 1 :
9
Input 2 :
3
1 6 5
Output 2 :
6
Whitelist
Set 1:
toString
equals

import java.util.*;

// Base class
class MaxFinder {

    // Method to find maximum value in array
    public int findMax(int[] arr) {
        int max = arr[0];

        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > max) {
                max = arr[i];
            }
        }

        return max;
    }
}

// Subclass extending MaxFinder
class ArrayData extends MaxFinder {

    private int[] values;

    // Constructor
    public ArrayData(int[] values) {
        this.values = values;
    }

    // Method to call parent class findMax
    public int findMax() {
        return super.findMax(values);
    }

    // Overriding toString() method
    public String toString() {
        return Arrays.toString(values);
    }

    // Overriding equals() method
    public boolean equals(Object obj) {

        if (this == obj)
            return true;

        if (obj == null || getClass() != obj.getClass())
            return false;

        ArrayData other = (ArrayData) obj;

        return Arrays.equals(this.values, other.values);
    }
}

// Main class
public class Main {

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        // Read number of elements
        int n = sc.nextInt();

        int[] values = new int[n];

        // Read array elements
        for (int i = 0; i < n; i++) {
            values[i] = sc.nextInt();
        }

        // Create object of subclass
        ArrayData data = new ArrayData(values);

        // Print maximum value
        System.out.println(data.findMax());

        sc.close();
    }
}
